### PR TITLE
refactor: project re-structured with src, utils and types

### PR DESCRIPTION
**Pull request by danielglobant**
# Summary

## The issue

refactor: project re-structured with src, utils and types #12

## Solution implemented

- Source code moved to a `src/` folder.
- `utils/` folder created and trasladed existing utils to this folder.
- `lib/` folder created with `types` and `interfaces`.
- `IRepositoryUseCase` interface created and used in corresponding uses cases for connecting with data repository.

## References
- https://github.com/Dylanflini/buscaminas-typescript/issues/8
- https://github.com/Dylanflini/buscaminas-typescript/issues/6

# Self review

- [ ] I have reviewed my PR changes and left comments/questions when necessary
- [ ] I have added a screenshot for the applied changes.
- [x] I have followed the style guidelines of this project.
- [x] I have added and ran the tests locally.
- [ ] I have ran the test coverage reports locally.
- [x] I have made corresponding changes to the documentation (README and API).
- [x] I have added the abstraction layers needed to separate responsibilities.

## Impact of these changes

This changes through `Component` impacts:

- `Component1`
- `Component2`
- `Component3`

## Evidences of expected changes

Add evidences here about each impact generated by your changes

# Type of change

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Hot fix (fixes an issue in master branch)
- [x] Docs (Documentation only changes)
- [ ] Style (Changes that do not affect the meaning of the code, formatting, missing semi-colons, etc)
- [x] Refactor (Changes that neither fixes a bug or adds a feature)
- [ ] Performance (Changes that improve performance)
- [ ] Chore (Changes to the build process or auxiliary tools and libraries)

# Squash and merge

We are following [Angular commit messages guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines)

```
<type>(<scope>): <Jira-Issue subject>
<BLANK LINE>
<optional body>
<BLANK LINE>
<optional footer>
```

Types

```
feat: A new feature
fix: A bug fix
docs: Documentation only changes
style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
refactor: A code change that neither fixes a bug nor adds a feature
perf: A code change that improves performance
test: Adding missing or correcting existing tests
chore: Changes to the build process or auxiliary tools and libraries such as documentation generation
```

Make sure you squash and merge your feature/bugfix branches into develop/release/hotfix branches using the suggested format.

Some examples.

```
feat: DP-1234 Edit assessment information
Frontend changes for editing an assessment:
- improve assign modal to receive an existing assessment
- integrate the reasign end point.
- couple of warnings removed
BREAKING CHANGE: the assign modal now receives 3 required parameters
```

More examples [here](https://www.conventionalcommits.org/en/v1.0.0-beta.2/#summary)

### CODE REVIEW

File: src/minesweeper/use-cases/startGame/startGame.ts [hunk 1]
```diff
@@ -24,36 +23,22 @@ type IBoardResponse = Omit<BoardModel, 'bombs' | 'neighBorsBombsCounter'>;
...
-    throw Error(ErrorStartGame.COLUMNS_GREATER_THAN_ZERO);
-  }
-
-  if (rows < 2 && columns < 2) {
-    throw new Error(ErrorStartGame.ROWS_AND_COLUMNS_GREATER_THAN_ONE);
-  }
+  if (bombsInput < 1) throw Error(ErrorStartGame.BOMBS_GREATER_THAN_ZERO);
+  if (rows < 1) throw Error(ErrorStartGame.ROWS_GREATER_THAN_ZERO);
+  if (columns < 1) throw Error(ErrorStartGame.COLUMNS_GREATER_THAN_ZERO);
+  if (rows < 2 && columns < 2) throw new Error(ErrorStartGame.ROWS_AND_COLUMNS_GREATER_THAN_ONE);
```
Comment (danielglobant) [inline src/minesweeper/use-cases/startGame/startGame.ts:37]: Not sure if we agree in this style of the `if`

Comment (Dylanflini) [reply src/minesweeper/use-cases/startGame/startGame.ts:37]: If is only for throw error it's ok

File: src/minesweeper/utils/position/position.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import { FlagModel } from '@minesweeper/domain/models';
+
+export const getSamePosition = (columnMarked: number, rowMarked: number) => {
+  const samePosition = ({ position: [column, row] }: IPosition) =>
+    column === columnMarked && row === rowMarked;
+
+  return samePosition;
+};
+
+export const getPosition = (column: number, row: number): FlagModel => ({
```
Comment (Dylanflini) [inline src/minesweeper/utils/position/position.ts:11]:
```suggestion
export const getPosition = (column: number, row: number): IPosition => ({
```

File: src/minesweeper/utils/position/position.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { IPosition } from '@minesweeper/lib/types';
+import { FlagModel } from '@minesweeper/domain/models';
+
+export const getSamePosition = (columnMarked: number, rowMarked: number) => {
```
Comment (Dylanflini) [inline src/minesweeper/utils/position/position.ts:4]: i think this function can get two parameters (position1: IPosition, position2: IPosition) and return a boolean

```suggestion
export const haveSamePosition = (firstPosition: IPosition, secondPosition: IPosition) => {
  const [fistX, firstY] = firstPosition
    const secondX, secondY] = secondPosition
  return  fistX === secondX && firstY === secondY;
```

File: src/minesweeper/use-cases/startGame/startGame.ts (hunk 1 above)
Comment (Dylanflini) [inline src/minesweeper/use-cases/startGame/startGame.ts:37]:
```suggestion
  if (rows < 2 && columns < 2) throw Error(ErrorStartGame.ROWS_AND_COLUMNS_GREATER_THAN_ONE);
```

**Pull request merged**